File: test-suite/src/test/java/io/micronaut/docs/inject/generics/Engine.java
package io.micronaut.docs.inject.generics;

// tag::class[]
public interface Engine<T extends CylinderProvider> { // <1>
    default int getCylinders() {
        return getCylinderProvider().getCylinders();
    }

    default String start() {
        return "Starting " + getCylinderProvider().getClass().getSimpleName();
    }

    T getCylinderProvider();
}
// tag::class[]

File: test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/inject/generics/Engine.kt
package io.micronaut.docs.inject.generics

// tag::class[]
interface Engine<T : CylinderProvider> { // <1>
    val cylinders: Int
        get() = cylinderProvider.cylinders

    fun start(): String {
        return "Starting ${cylinderProvider.javaClass.simpleName}"
    }

    val cylinderProvider: T
}
// end::class[]